File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/util/AnimationUtil.java
package com.dbottillo.mtgsearchfree.util;

import android.animation.ArgbEvaluator;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;
import android.view.View;

public final class AnimationUtil {

    private AnimationUtil() {

    }

    private static final int DEFAULT_DURATION = 200;

    public static ValueAnimator animateHeight(final View view, int target) {
        if (view.getHeight() == target) {
            return null;
        }
        ValueAnimator anim = ValueAnimator.ofInt(view.getHeight(), target);
        anim.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                int val = (Integer) valueAnimator.getAnimatedValue();
                UIUtil.setHeight(view, val);
            }
        });
        anim.setDuration(DEFAULT_DURATION);
        anim.start();
        return anim;
    }

    public static void growView(View view) {
        view.setScaleX(0.0f);
        view.setScaleY(0.0f);
        ObjectAnimator scaleUp = ObjectAnimator.ofPropertyValuesHolder(view,
                PropertyValuesHolder.ofFloat("scaleX", 1.0f),
                PropertyValuesHolder.ofFloat("scaleY", 1.0f));
        scaleUp.setDuration(250);
        scaleUp.start();
    }

    public static void shrinkView(View view) {
        ObjectAnimator scaleUp = ObjectAnimator.ofPropertyValuesHolder(view,
                PropertyValuesHolder.ofFloat("scaleX", 0.0f),
                PropertyValuesHolder.ofFloat("scaleY", 0.0f));
        scaleUp.setDuration(250);
        scaleUp.start();
    }

    public static LinearInterpolator createLinearInterpolator() {
        return new LinearInterpolator();
    }

    public static class LinearInterpolator implements TimeInterpolator {

        private float mStartValue;
        private float mEndValue;

        LinearInterpolator() {
        }

        public final LinearInterpolator fromValue(final float startValue) {
            mStartValue = startValue;
            return this;
        }

        public final LinearInterpolator toValue(final float endValue) {
            mEndValue = endValue;
            return this;
        }

        @Override
        public final float getInterpolation(final float input) {
            final float amount = Math.abs(mEndValue - mStartValue);
            if (mEndValue > mStartValue) {
                return mStartValue + (amount * input);
            } else {
                return mStartValue - (amount * input);
            }
        }
    }

    public static ArgbInterpolator createArgbInterpolator() {
        return new ArgbInterpolator();
    }

    public static class ArgbInterpolator {

        private int mStartValue;
        private int mEndValue;
        private ArgbEvaluator argbEvaluator;

        ArgbInterpolator() {
            argbEvaluator = new ArgbEvaluator();
        }

        public final ArgbInterpolator fromValue(final int startValue) {
            mStartValue = startValue;
            return this;
        }

        public final ArgbInterpolator toValue(final int endValue) {
            mEndValue = endValue;
            return this;
        }

        public final int getInterpolation(final float input) {
            return (int) argbEvaluator.evaluate(input, mStartValue, mEndValue);
        }
    }

}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/util/AnimationUtil.kt
package com.dbottillo.mtgsearchfree.util

import android.animation.*
import android.view.View

object AnimationUtil {

    private val DEFAULT_DURATION = 200

    fun animateHeight(view: View, target: Int): ValueAnimator? {
        if (view.height == target) {
            return null
        }
        val anim = ValueAnimator.ofInt(view.height, target)
        anim.addUpdateListener { valueAnimator ->
            val `val` = valueAnimator.animatedValue as Int
            view.setHeight(`val`)
        }
        anim.duration = DEFAULT_DURATION.toLong()
        anim.start()
        return anim
    }

    fun growView(view: View) {
        view.scaleX = 0.0f
        view.scaleY = 0.0f
        val scaleUp = ObjectAnimator.ofPropertyValuesHolder(view,
                PropertyValuesHolder.ofFloat("scaleX", 1.0f),
                PropertyValuesHolder.ofFloat("scaleY", 1.0f))
        scaleUp.duration = 250
        scaleUp.start()
    }

    fun shrinkView(view: View) {
        val scaleUp = ObjectAnimator.ofPropertyValuesHolder(view,
                PropertyValuesHolder.ofFloat("scaleX", 0.0f),
                PropertyValuesHolder.ofFloat("scaleY", 0.0f))
        scaleUp.duration = 250
        scaleUp.start()
    }

    fun createLinearInterpolator(): LinearInterpolator {
        return LinearInterpolator()
    }

    class LinearInterpolator : TimeInterpolator {

        private var mStartValue: Float = 0.toFloat()
        private var mEndValue: Float = 0.toFloat()

        fun fromValue(startValue: Float): LinearInterpolator {
            mStartValue = startValue
            return this
        }

        fun toValue(endValue: Float): LinearInterpolator {
            mEndValue = endValue
            return this
        }

        override fun getInterpolation(input: Float): Float {
            val amount = Math.abs(mEndValue - mStartValue)
            return if (mEndValue > mStartValue) {
                mStartValue + amount * input
            } else {
                mStartValue - amount * input
            }
        }
    }

    fun createArgbInterpolator(): ArgbInterpolator {
        return ArgbInterpolator()
    }

    class ArgbInterpolator {

        private var mStartValue: Int = 0
        private var mEndValue: Int = 0
        private val argbEvaluator: ArgbEvaluator = ArgbEvaluator()

        fun fromValue(startValue: Int): ArgbInterpolator {
            mStartValue = startValue
            return this
        }

        fun toValue(endValue: Int): ArgbInterpolator {
            mEndValue = endValue
            return this
        }

        fun getInterpolation(input: Float): Int {
            return argbEvaluator.evaluate(input, mStartValue, mEndValue) as Int
        }
    }

}
